File: node/src/providers/model-catalog.ts
import type { ModelRegistration } from "./model-registry.js"

export type { ModelRegistration } from "./model-registry.js"

/** A query-only model catalog. Catalogs describe models; selection policy belongs to CapabilityRouter. */
export interface ModelCatalog {
  list(): Promise<readonly ModelRegistration[]>
  get(modelId: string): Promise<ModelRegistration | undefined>
}

export interface ModelCatalogSource {
  list(): Promise<readonly ModelRegistration[]>
}

export type ModelCatalogRefreshResult =
  | { ok: true }
  | { ok: false; errorCode: "refresh_failed" }

/** Immutable, deterministic model facts supplied by the SDK or an application. */
export class StaticModelCatalog implements ModelCatalog {
  private readonly registrations: readonly ModelRegistration[]
  private readonly byId: ReadonlyMap<string, ModelRegistration>

  constructor(registrations: readonly ModelRegistration[]) {
    const byId = new Map<string, ModelRegistration>()
    for (const registration of registrations) {
      const id = registration.descriptor.id
      if (byId.has(id)) throw new Error(`Duplicate model catalog entry: ${id}`)
      byId.set(id, registration)
    }
    this.registrations = [...byId.values()].sort((a, b) => a.descriptor.id.localeCompare(b.descriptor.id))
    this.byId = byId
  }

  async list(): Promise<readonly ModelRegistration[]> {
    return this.registrations
  }

  async get(modelId: string): Promise<ModelRegistration | undefined> {
    return this.byId.get(modelId)
  }
}

/**
 * A host-owned discovery cache. Refresh never throws and never deletes the last good snapshot;
 * callers can keep using the static catalog when a remote provider is unavailable.
 */
export class DynamicModelCatalog implements ModelCatalog {
  private snapshot = new Map<string, ModelRegistration>()

  constructor(
    private readonly source: ModelCatalogSource,
    private readonly fallback: ModelCatalog = new StaticModelCatalog([]),
  ) {}

  async list(): Promise<readonly ModelRegistration[]> {
    const merged = new Map<string, ModelRegistration>()
    for (const registration of await this.fallback.list()) merged.set(registration.descriptor.id, registration)
    for (const registration of this.snapshot.values()) merged.set(registration.descriptor.id, registration)
    return [...merged.values()].sort((a, b) => a.descriptor.id.localeCompare(b.descriptor.id))
  }

  async get(modelId: string): Promise<ModelRegistration | undefined> {
    return this.snapshot.get(modelId) ?? await this.fallback.get(modelId)
  }

  async refresh(): Promise<ModelCatalogRefreshResult> {
    try {
      const next = new Map<string, ModelRegistration>()
      for (const registration of await this.source.list()) {
        const id = registration.descriptor.id
        if (next.has(id)) throw new Error("duplicate dynamic model catalog entry")
        next.set(id, registration)
      }
      this.snapshot = next
      return { ok: true }
    } catch {
      return { ok: false, errorCode: "refresh_failed" }
    }
  }
}

File: python/deepstrike/providers/model_catalog.py
"""Static model facts and host-owned dynamic discovery cache.

Catalogs describe available models only. Capability selection remains in
``capability_router`` so discovery cannot become an implicit routing policy.
"""
from __future__ import annotations

from typing import Protocol

from .model_registry import ModelRegistration


class ModelCatalog(Protocol):
    async def list(self) -> tuple[ModelRegistration, ...]: ...
    async def get(self, model_id: str) -> ModelRegistration | None: ...


class ModelCatalogSource(Protocol):
    async def list(self) -> tuple[ModelRegistration, ...] | list[ModelRegistration]: ...


class StaticModelCatalog:
    """Immutable deterministic facts supplied by the SDK or application."""

    def __init__(self, registrations: tuple[ModelRegistration, ...] | list[ModelRegistration]) -> None:
        by_id: dict[str, ModelRegistration] = {}
        for registration in registrations:
            identifier = registration.descriptor.id
            if identifier in by_id:
                raise ValueError(f"Duplicate model catalog entry: {identifier}")
            by_id[identifier] = registration
        self._by_id = by_id
        self._registrations = tuple(by_id[key] for key in sorted(by_id))

    async def list(self) -> tuple[ModelRegistration, ...]:
        return self._registrations

    async def get(self, model_id: str) -> ModelRegistration | None:
        return self._by_id.get(model_id)


class DynamicModelCatalog:
    """Refreshable discovery snapshot with static fallback and last-good retention."""

    def __init__(self, source: ModelCatalogSource, fallback: ModelCatalog | None = None) -> None:
        self._source = source
        self._fallback = fallback or StaticModelCatalog(())
        self._snapshot: dict[str, ModelRegistration] = {}

    async def list(self) -> tuple[ModelRegistration, ...]:
        merged = {item.descriptor.id: item for item in await self._fallback.list()}
        merged.update(self._snapshot)
        return tuple(merged[key] for key in sorted(merged))

    async def get(self, model_id: str) -> ModelRegistration | None:
        return self._snapshot.get(model_id) or await self._fallback.get(model_id)

    async def refresh(self) -> dict[str, object]:
        try:
            next_snapshot: dict[str, ModelRegistration] = {}
            for registration in await self._source.list():
                identifier = registration.descriptor.id
                if identifier in next_snapshot:
                    raise ValueError("duplicate dynamic model catalog entry")
                next_snapshot[identifier] = registration
            self._snapshot = next_snapshot
            return {"ok": True}
        except Exception:
            return {"ok": False, "error_code": "refresh_failed"}
